### ISSUE TITLE
MSVC taking more than 1 CPU

### ISSUE THREAD

**Issue by positivelink** [fastbuild/fastbuild#521]
I have seen cases of cl.exe (MSVC 14.13.26128) taking more than 1 CPU (up to 4) when compiling, defeating the 1-process-per-core heuristic.
This is consistent with what is documented in https://docs.microsoft.com/en-us/cpp/build/reference/cgthreads-code-generation-threads?view=vs-2017, and adding `/cgthreads1` to a command line that did use more than 1 core seemed to address the issue.

I suggest adding the flag by default to MSVC command lines, is `ObjectNode::BuildArgs()` the right place to do it ?

**Comment by ffulin** [comment]
cl.exe can indeed use more than one CPU core (as can link.exe). While there are many situations where it would be better if it only used one CPU core per invocation for overall performance, there are also situations where having it using slightly more cores per physical core is beneficial. Specifically:
 - rapidly iterating on small changes (compiling < NUM_PROCESSORS files)
 - at the tail end of a build (for the last few tasks where we go under < NUM_PROCESSORS)
 - if there are other bottlenecks in the build that result in under utilization
 - in network or IO bound scenarios where cores can also be otherwise underutilized

How much the default cgthreads helps in some cases and hurts in others likely varies between codebases. Given that there is a tradeoff here, one situation is not necessarily better (or more important to a given user), and there isn't a one-size fits all solution, I don't think it's right that FASTBuild should try to decide what is better and modify the default behaviour. Even if it tried decide dynamically at every process invocation, it would just as often make things worse as it would make them better. Additionally, if a given user decides for their favor one thing over the other, they can set /cgthreads themselves in the bff.

Do you think /cgthreads might be better for your use case? If you didn't already, perhaps you can profile it, and if you like it better overall, you could add it to your command line in the bff.

Let me know what you think.

**Comment by positivelink** [comment]
The case where it hurts most is actually with FBuildWorker, when my machine compiles remote jobs - depending on the context, builds may take much more CPU than what FBuildWorker is configured to use.
I understand how a blanket change can actually have more negative impact than positve. I'll run a benchmark and get back to you with more information.

**Comment by ffulin** [comment]
**> The case where it hurts most is actually with FBuildWorker,...**

Ah I see. I could potentially get on board with overriding the behaviour in a case like this if we can prove out a better behaviour overall. I do suspect there is a somewhat similar tradeoff though - perhaps there is a world where the remote worker could apply some sensible logic to decide when to do this or not.

One tricky thing though is that /cgthreads is a relatively new option, only present in more recent cl.exe. FATSBuild doesn't know what version of cl.exe we are running, so that might be something we'd have to introduce.

If you gather some data, please let me know and we can see how to proceed.

**Comment by bburgin** [comment]
You could limit older as well as newer cl.exe's to 1 cpu, if you add an option to the worker that causes it to spawn child processes with a JOB_OBJECT_LIMIT_AFFINITY.
https://docs.microsoft.com/en-us/windows/desktop/api/winnt/ns-winnt-_jobobject_basic_limit_information
This way, you don't have to rely on the cl.exe /cgthreads being there.

You could instead use the sandbox feature I made a pull request for.
Launch the worker with -sandboxexe=winc.exe and limit the number of cores by passing in a -sandboxargs=--affinity 1
https://github.com/iceb0y/windows-container

**Comment by ffulin** [comment]
@bburgin Thanks for that link - there is a lot of interesting stuff there. I took a look and eventually ended up looking at some info that talks about processor percentage limits for Job Objects: https://docs.microsoft.com/en-us/windows/desktop/api/winnt/ns-winnt-_jobobject_cpu_rate_control_information

I actually wonder if this might be a good way to solve this problem. If we could limit FBuildWorker and child processes to an overall scheduling percentage, then it wouldn't matter how the threads were set up and we also wouldn't need to mess around with processor affinity (which can be problematic and complicated, especially on systems with >64 CPUs).

This probably needs some more investigation, but it seems like an interesting route to look into.

**Comment by bburgin** [comment]
I agree. JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP looks like the best option. It seems like it would allow the OS to spread a cl.exe's thread work across multiple cpus, if needed; but limit the cl.exe's overall cpu usage to a percentage of the machine's total cpu capacity. If so, this would be the most performant approach.

I found this example code:
https://github.com/goldshtn/jobrun

Alternatively, we could add a cpu rate option to winc.exe using the same JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP flag.

**Comment by bburgin** [comment]
I found some more info in the comment section of this post:
https://stackoverflow.com/a/9793403

The JOB_OBJECT_CPU_RATE api is Win8 and higher.
I am still using Windows 7 for some machines.
So consider using NtSetInformationProcess and QUOTA_LIMITS_EX::CPU_RATE_LIMIT instead.

CPU rate does not affect I/O throughput, but process priority does.
Are you using process priority for child processes in FBuild or FBuildWorker?
If so, I am curious if I/O throughput is lower because of this?

Would performance data would help steer further discussion?

**Comment by ffulin** [comment]
FBuildWorker and FBuild both set themselves to BELOW_NORMAL priority at startup (Windows only), which is inherited by all child processes:
```C++
SetPriorityClass( GetCurrentProcess(), BELOW_NORMAL_PRIORITY_CLASS );
```

The reasoning is that we don't want the overall usability of the machine to be negatively impacted. My understanding of this priority mechanism is that it only impacts performance when there are higher priority processes.

It's possible a better mechanism for controlling priority that either complements or replaces this would be better (as we're talking about now).

The reasoning behind this lowering of priority is as follows:

**FBuildWorker** 

For the FBuildWorker case, when an "idle" machine is working for others, and the user starts something to leave the idle state, we want FBuildWorker and its child processes to be de-prioritized (CPU and IO) so use of the machine favors interactivity and user-spawned processes. Ideally, FBuildWorker would actually cancel in-progress tasks (something we can probably support now that we have process cancellation support implemented in Process.cpp). The Idle detection already has de-bouncing that should avoid problematic edge-cases transitioning states.

Users get upset if they think the shared worker is making their machine run poorly, and we don't want them cancelling or closing FBuildWorker and reducing the pool of workers, so this de-prioritization is important.

If we can limit overall CPU usage in a better way as we've been talking about, I see that as complimentary to this behaviour rather than as a replacement, but I'm open to controlling the priority in a better way if there is one.

In cases where FBuildWorker is used on a dedicated box (and presumably running in dedicated {"Work For Others Always"} mode) I could see an argument for not lowering priority this way, but at the same time, if the box is dedicated, nothing else should be going on anyway and in that case the lower priority should have no impact on performance anyway.

**FBuild**

FBuild sets itself to BELOW_NORMAL priority in order to keep the machine in a usable state - specifically to ensure interactive applications remain interactive. Without this, many machines become pretty much unusable. This option helped ensure UI applications (like Visual Studio) remain responsive for example.

If the machine is doing nothing else, overall FBuild performance is not impacted in any way that I've seen. If the user is doing other stuff, I think it's correct to prioritize those interactive applications about compilation speed. People get upset when that can't use the internet while compiling :)

There are some unfortunate cases I've seen - things like Visual Assist deciding that the best time to rebuild its database is when a compilation starts. In this case, FBuild is lower priority than this scanning, but there is no sustainable way for us to know this and in this case the thing that needs fixing is the poor behaviour of Visual Assist.

Windows 10 seems to have a lot of process scheduling and interactivity problems load (many of which are documented in blog posts by Bruce Dawson), so some machines still suffer from poor usability when under load without much apparent rhyme or reason.  My work PC has mouse hangs and stall when under load in Windows 10 (not just with FBuild) but my home PC does not, despite both being the same Windows 10 version.

As with the FBuildWorker, I'm open to using complimentary or better APIs to achieve the desired effect of de-prioritization vs user-interactive tasks.

**Comment by bburgin** [comment]
Sounds good. Thanks.
